spring-integration-event/src/main/java/org/springframework/integration/event/inbound/ApplicationEventListeningMessageProducer.java

```diff
@@ -116,20 +129,45 @@ public void onApplicationEvent(ApplicationEvent event) {
 		}
 	}
 
+	@Override
+	protected Object evaluatePayloadExpression(Object rootObject) {
+		if (!this.payloadExpressionSet && rootObject instanceof PayloadApplicationEvent) {
+			return ((PayloadApplicationEvent) rootObject).getPayload();
```

Comment:
> I don't think we expect users to publish arbitrary object as events.

You probably should as it was a requested feature. To be clear, it's still an event, it's just that it does not extend from our `ApplicationEvent`. So, it still has some kind of structure but in practice it _is_ an arbitrary event. `PayloadApplicationEvent` is just a way for us to transmit such object as a regular event so that the (internal) API doesn't have to change for this new use case.